boot/bootstate20.go

```diff
@@ -356,31 +356,96 @@ func (envbks *envRefExtractedKernelBootloaderKernelState) setNextKernel(sn snap.
 }
 
 //
-// kernel snap methods
+// bootStateUpdate for 20 methods
 //
 
-// bootState20Kernel implements the bootState and bootStateUpdate interfaces for
-// kernel snaps on UC20. It is used for setNext() and markSuccessful() - though
-// note that for markSuccessful() a different bootStateUpdate implementation is
-// returned, see bootState20MarkSuccessful
-type bootState20Kernel struct {
-	bks bootloaderKernelState20
+type bootCommitTask func() error
+
+// bootState20Kernel implements the bootStateUpdate interface for both kernel
+// and base snaps on UC20.
+type bootStateUpdate20 struct {
+	// tasks to run before the modeenv has been written
+	preModeenvTasks []bootCommitTask
+
+	// the modeenv that was read from the state
+	modeenv *Modeenv
 
-	// the kernel snap that was booted for markSuccessful()
-	bootedKernelSnap snap.PlaceInfo
+	// the modeenv that will be written out in commit
+	writeModeenv *Modeenv
 
-	// the kernel snap to try for setNext()
-	nextKernelSnap snap.PlaceInfo
+	// tasks to run after the modeenv has been written
+	postModeenvTasks []bootCommitTask
+}
```

Comment:
it's a bit more code but a couple of helper methods preModeenv(task),  postModeenv(task) to append to the list would be good

Reply:
sure